# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def addTwo(x):
    return x + 2

print(addTwo(4))

# Lambda Function Syntax
# Example of calling a lambda function without a variable
print((lambda x: x + 2)(4))

6
6


#### Saving to a Variable

In [3]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [4]:
# Multiple Inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

# Multiple Inputs with a variable attached
x_test = lambda x,y,z: z - y - x
print(x_test(3,5,8))

120
0


#### Passing a Lambda into a Function

In [6]:
def multiply(f,num):
    """
        f expects a lambda function
        num expects a number
    """
    return f(num)
multiply(lambda x: x * x, 4) # lambda 4: 4 * 4
# f = lambda x: x*x

16

#### Returning a Lambda from a Function

In [18]:
# Regular defined function
def multiply_test(num):
    return num * 4

# Function within a function
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x,a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x00000281F87A3EE8>
8
19


#### If Statements within Lambdas

In [22]:
# Lambda x : True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))

print(f_condition(12))

print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [7]:
f = lambda num: num ** 3
print(f(4))
print(f(5))

64
125


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [77]:
#map() way
def square(x):
    return x ** 2
output = list(map(square, [1,2,3,4,5]))
print(output)



# regular for loop + append
def regWay(list1):
    output2 = []
    for num in list1:
        output2.append(num**2)
    return output
print(regWay([1,2,3,4,5]))



# list comprehension
def lcWay(list1):
    return[num**2 for num in list1]
print(lcWay([1,2,3,4,5]))

[1, 4, 9, 16, 25]
[1, 4, 9, 16, 25]
[1, 4, 9, 16, 25]


In [28]:
# map(func,iterable(list,dict,tuple,set))
# Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
def squared(num,num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else:
        return num,num2
    
numbers = [4,11,20,3,15,20]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared,numbers,more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [29]:
# map(lambda x: x + 2, list)
# using lambda in map happens in line (or in one line) usually
# list(map(lambda x : x * 2,nums))
squared_nums_lamb = list(map(lambda x,y : (x ** 2,y ** 2) if x < 10 and y < 10 else (x,y),numbers,more_nums))
print(squared_nums_lamb)


[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


In [30]:
list(map(lambda x: x**2 if x<5 else x, [4,5,6,7,8]))
# 4 is less than 5 so it is squared

[16, 5, 6, 7, 8]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [36]:
listOfNums = [1,2,3,4,5]
list(map(lambda num: num*2-1, listOfNums))

[1, 3, 5, 7, 9]

In [51]:
double_num = list(map(lambda x,y: (((x * 2) - 1),((y * 2) - 1)) if x > 0 and y > 0 else (x,y), numbers,more_nums))
print(double_num)

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [38]:
# map(func, iter)   map adds the return value of func() to our new list
# filter(function, iterable)  filter adds the same value from iterable, based on whether the return value of func() is True or not.


#filter numbers that are even
def filterOutEven(num):
    if num % 2 == 0:  # if number is even, then return False value
        return False
    return True
    
list(filter(filterOutEven, [1,2,3,4,5,6,7,8,9]))
# BECAUSE 1 isn't an even number, it gets returned TRUE. so 1 is in the output list. ONLY TRUE ELEMENTS ARE ADDED TO THE LIST.

[1, 3, 5, 7, 9]

In [10]:
names = ['Bob','Andy','Max','Evan','Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)

print(a_names)

['Andy', 'Angelica']
<function a_names at 0x00000161458301F0>


#### Using Lambda's with Filter()

In [39]:
list(filter(lambda num: False if num % 2 == 0 else True, [1,2,3,4,5,6,7,8,9]  ))

[1, 3, 5, 7, 9]

In [57]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [55]:
import statistics

themean = statistics.mean([1,2,3,4,5,6,7,8,9])

print(list(filter(lambda number: False if number < themean else True, [1,2,3,4,5,6,7,8,9])))

[5, 6, 7, 8, 9]


In [62]:
from statistics import mean

nums = [2,7,4.2,1.6,9,4.4,4.9]

print(mean(nums))

avg = mean(nums)
print(avg)

def mean_test(name):
    if name < avg:
        return True
    else:
        return False
new_calc = list(filter(mean_test,nums))
print(new_calc)

# Lambda function solution
below_avg = list(filter(lambda x: True if x <= mean(nums) else False, nums))
print(below_avg)

4.728571428571429
4.728571428571429
[2, 4.2, 1.6, 4.4]
[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [67]:
from functools import reduce

# reduce(function,iterable)

list_1 = [2,4,6,8,10]

def addNums(num1,num2):
    return num1 + num2

result_add = reduce(addNums,list_1)

print(result_add)

# Subtract a list of numbers
def subtractNums(num1,num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)


30
-26


In [58]:
from functools import reduce
list1 = [3]
def addNums(num1,num2):
    print('num1', num1)
    print('num2', num2)
    return num1 + num2

result = reduce(addNums, list1)
print(result)

3


#### Using Lambda's with Reduce()

In [68]:
result_lamb = reduce(lambda x,y: x + y, list_1)   # lambda  : return statement, iterable
print(result_lamb)

30


NameError: name 'reduce' is not defined

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [59]:
reduce(lambda x,y: x * y, [1,2,3,4,5])

120

In [60]:
reduce(lambda x,y: x*y if x%2 == 0 else x//y, [6,2,3,4,5])


720

In [69]:
my_list = [1,2,3,4]

from functools import reduce

result_lamb = reduce(lambda x,y: x * y, my_list)
print(result_lamb)


24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [1]:
def addNums(num):
    # Set base case for recusive function
    if num <= 1:
        print("addNums(1) = 1")
        return num  # output is 15
    else:
        print(f"addNums({num}) = {num} + addNums({num - 1})")
        return num + addNums(num - 1)
    
addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

In [62]:
def infinite(num):
    if num == 1:
        return "done"
    print(f'{num}, looping again')
    return infinite(num-1)
infinite(10)

10, looping again
9, looping again
8, looping again
7, looping again
6, looping again
5, looping again
4, looping again
3, looping again
2, looping again


'done'

#### Writing a Factorial Function

In [61]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1) # for first one is 5 *4. next loop is 4*3
    
factorial(5)


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [66]:
# return num - subnum(num-1)

def subnum(num):
    if num<=1:
        return 1
    print(f'{num} = {num} - subnum({num - 1}) ')
    return num - subnum(num - 1)

subnum(15)

15 = 15 - subnum(14) 
14 = 14 - subnum(13) 
13 = 13 - subnum(12) 
12 = 12 - subnum(11) 
11 = 11 - subnum(10) 
10 = 10 - subnum(9) 
9 = 9 - subnum(8) 
8 = 8 - subnum(7) 
7 = 7 - subnum(6) 
6 = 6 - subnum(5) 
5 = 5 - subnum(4) 
4 = 4 - subnum(3) 
3 = 3 - subnum(2) 
2 = 2 - subnum(1) 


8

In [5]:
# return num - subNums(num - 1)
# subtract(10)
def subNums(num):
    if num <= 1:
        print("subNums(1) = 1")
        return num
    else:
        print(f"subNums({num}) = {num} - subNums({num - 1})")
        return num - subNums(num - 1)
    
subNums(10)

subNums(10) = 10 - subNums(9)
subNums(9) = 9 - subNums(8)
subNums(8) = 8 - subNums(7)
subNums(7) = 7 - subNums(6)
subNums(6) = 6 - subNums(5)
subNums(5) = 5 - subNums(4)
subNums(4) = 4 - subNums(3)
subNums(3) = 3 - subNums(2)
subNums(2) = 2 - subNums(1)
subNums(1) = 1


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [69]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [72]:
def myRange(start, stop, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step
myRange(2,20)
for i in myRange(2,20):
    myGeneratorValue = i
    print(i)
    
print(myGeneratorValue)

2
4
6
8
10
12
14
16
18
18


In [14]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step
for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)

my_range(20,start = 2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x000002EB1C6E5C48>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [76]:
def myGen(num):
    while num > 0:
        yield num ** 2
        num -= 1 #shorthand for num = num - 1
        
for x in myGen(10):
    print(x)
print(list(myGen(10)))

100
81
64
49
36
25
16
9
4
1
[100, 81, 64, 49, 36, 25, 16, 9, 4, 1]


In [74]:
# for i in numbers_squared(10)

def num_sq(stop,start,step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in num_sq(0,start = 3):
    print(i)

9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [112]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]


places = filter(lambda name: name.strip(), places)
list(places)

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Andrew P. Garfield', 'David hassELHOFF', 'Shoha Tsuchida']`

In [149]:
author = ["Shoha Tsuchida", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

ordered = sorted(author, key=lambda s: s.split()[-1].casefold())
print(ordered)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Andrew P. Garfield', 'David hassELHOFF', 'Shoha Tsuchida']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [142]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

cities = ['Nasua', 'Boston', 'Los Angelos', 'Miami']
celsius = [32,12,44,29]

fahrenheit = list(map(lambda x,y: y*(9/5) + 32 else (x,y) cities, celsius))

SyntaxError: invalid syntax (<ipython-input-142-6ab823b3cc61>, line 7)

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [25]:
def fib(num):
    if num<=1:
        return num
    return (fib(num-1) + fib(num - 2))

fibseq = 7

if fibseq > 0:
    print("fibonacci sequence")
    for i in range(fibseq):
        print(fib(i))

fibonacci sequence
0
1
1
2
3
5
8
